### PR TITLE
Fix MIDI player tempo reset issues

### PR DESCRIPTION
**Pull request by jjceresa**
This PR addresses issue https://github.com/FluidSynth/fluidsynth/issues/707 according to the last discussion.
@glm35, please, your are welcome to pull this branch `fix-player-tempo` an test the new function `fluid_player_set_tempo()`, and `fluid_player_get_tempo()`.

### CODE REVIEW

**Comment by jjceresa** [comment]
Please, as I am not familiar with doxygen formatting rule, could you have a check , in particular for `fluid_player_tempo_type` enum in `midi.h` file ?.

File: include/fluidsynth/midi.h [hunk 1]
```diff
@@ -193,7 +193,7 @@ FLUIDSYNTH_API void delete_fluid_midi_driver(fluid_midi_driver_t *driver);
 /* @} */
 
 /**
- * @defgroup midi_player MIDI Player
+ * @defgroup midi_player MIDI file Player
```
Comment (mawe42) [inline include/fluidsynth/midi.h:196]: Nitpick: for places where we use title case (in the name of groups, for example), we should ensure to use it consistently. So use "MIDI File Player" here, not "MIDI file Player".

**Comment by derselbst** [comment]
Looks good to me. I'll keep this open for a few days to give @glm35 a chance to test this.

**Comment by jjceresa** [comment]
>Looks good to me....

Ok. Thanks for reviewing and fixing API docs, and grammar.

**Comment by glm35** [comment]
Thanks for this PR!

I did some basic functional tests:
- midi file with tempo and midi file without tempo
- with and without calling set_tempo(FLUID_PLAYER_TEMPO_BPM flavor) before playback start
- file played one time, two times in a playlist, three times in a loop

All good for the moment, and this solves all the problems I reported in issue #707.

I intend to do more tests: set tempo after playback start, get_tempo(), midi file with tempo change, FLUID_PLAYER_TEMPO_MIDI, FLUID_PLAYER_TEMPO_DEFAULT and FLUID_PLAYER_TEMPO_RELATIVE. It will probably take me the rest of the week to find some time to do that.

**Comment by derselbst** [comment]
Great, thanks for having a first try! And no problem, take your time.

**Comment by glm35** [comment]
I did a few more tests today:
- set tempo 3s after playback start, with midi file with and without "set tempo"
- midi file with tempo change

Still all good :-)

... but test plan not completed yet...

**Comment by derselbst** [comment]
> ... but test plan not completed yet...

@glm35 Have you had opportunity to complete you test plan successfully?

**Comment by glm35** [comment]
No, I haven't.

I took a side step to build a small player GUI that will also help me for the tempo API tests (https://github.com/glm35/pyfluidsynth3/blob/gl-devel-new-tempo-api/examples/player_gui.py).

I will have more time to work on this this week, so if you can wait a litter more...

**Comment by derselbst** [comment]
> I will have more time to work on this this week, so if you can wait a litter more...

Sure, no problem.

**Comment by glm35** [comment]
An update about my testing of the new tempo API...

I tested fluid_player_set_tempo() with FLUID_PLAYER_TEMPO_DEFAULT, and it is fine for me. When the midi file contains "set tempo" messages, the playback continues with the last value seen in the file, which is nice. When the midi file does not contain any "set tempo", the playback continues with fluidplayer default tempo (120 bpm), which what I expected.

**Comment by glm35** [comment]
I also tested fluid_player_get_tempo(). For me, it is working as documented. However, I find this new API a bit complex to understand and to use.

I think that the main use case for this API is to get the current tempo used by the player when the tempo values come from the the midi file. If you want to know this, you need to call get_tempo() with FLUID_PLAYER_TEMPO_DEFAULT, and I find this confusing because from me the default tempo value is the value used by the player when no tempo is specified by the user or in the midi file (120 bpm). Maybe it would be clearer to rename FLUID_PLAYER_TEMPO_DEFAULT to FLUID_PLAYER_TEMPO_MIDI_FILE? Also, you will only get the tempo value in "midi tempo" format (us/quarter note). If you want it in bpm, you need to compute it.

Calling get_tempo() with FLUID_PLAYER_TEMPO_BPM or with FLUID_PLAYER_TEMPO_MIDI will give you the value configured by the user with set_tempo() if any, else the default value used by the player (120 bpm). This is consistent with set_tempo(), but not very useful IMO. I would prefer that calling set_tempo() with those tempo types would return the current tempo value in the desired format (bpm or midi). I think this would make the API more intuitive to use, and also it would allow the user to get the current tempo in bpm.

To conclude, beyond troubleshooting, I personally don't have a real use case for get_tempo(). But I think that it might be enough to have a simpler API that would just return the current tempo used by the player in the desired format, ie that would only accept the tempo types FLUID_PLAYER_TEMPO_BPM ans FLUID_PLAYER_TEMPO_MIDI, and that would drop the sync_mode parameter. But that is just my point of view.

File: src/midi/fluid_midi.c [hunk 2]
```diff
@@ -2232,23 +2243,142 @@ int fluid_player_set_loop(fluid_player_t *player, int loop)
...
+ * change or controlled by external tempo expressed in BPM or in micro seconds
+ * per quarter note.
+ *
+ * @param player MIDI player instance. Must be a valid pointer.
+ * @param tempo_type Must a be value of #fluid_player_tempo_type and indicates the
+ *  meaning of tempo value and how the player will be controlled, see below.
+ * @param tempo Tempo value or multiplier.
+ * 
+ *  #FLUID_PLAYER_TEMPO_DEFAULT, the player will be controlled by internal MIDI
+ *  file tempo changes. The @c tempo parameter will be ignored.
```
Comment (glm35) [inline src/midi/fluid_midi.c:2291]: Could you precise that if no tempo is given in the midi file, a default value of 120 bpm will be used?

File: src/midi/fluid_midi.c [hunk 3]
```diff
@@ -2232,23 +2243,142 @@ int fluid_player_set_loop(fluid_player_t *player, int loop)
...
+ *
+ *  #FLUID_PLAYER_TEMPO_BPM, the player will be controlled by the external tempo
+ *  value provided  by the tempo parameter in bpm (i.e in quarter notes per minute)
+ *  which must be in the range (1 to 60000000).
```
Comment (glm35) [inline src/midi/fluid_midi.c:2295]: 60000000 bpm, really?

File: src/midi/fluid_midi.c [hunk 4]
```diff
@@ -1608,7 +1609,9 @@ fluid_track_send_events(fluid_track_t *track,
 
         if(event->type == MIDI_SET_TEMPO && player != NULL)
         {
-            fluid_player_set_midi_tempo(player, event->param1);
+            /* memorize the tempo change value comming from the MIDI file */
```
Comment (glm35) [inline src/midi/fluid_midi.c:1612]: typo: comming -> coming
(can also be found in other places, I only report this one)

File: src/midi/fluid_midi.c [hunk 5]
```diff
@@ -2317,10 +2448,98 @@ int fluid_player_get_total_ticks(fluid_player_t *player)
...
+ * per quarter note.
+ *
+ * @param player MIDI player instance. Must be a valid pointer.
+ * @param tempo_type A value of #fluid_player_tempo_type that indicates the
+ * tempo information to retrieve, see below.
+ *
+ * @param tempo Pointer to the value the returned information will be written to.
+ *  This pointer can be NULL to ignore the tempo information
+ *
+ * @param sync_mode Pointer an int for returning the tempo mode the player
```
Comment (glm35) [inline src/midi/fluid_midi.c:2464]: typo: "Pointer an int" -> "Pointer to an int"

**Comment by jjceresa** [comment]
>Maybe it would be clearer to rename FLUID_PLAYER_TEMPO_DEFAULT to FLUID_PLAYER_TEMPO_MIDI_FILE? 

Basically you right. This name is confusing. The enum name should be FLUID_PLAYER_TEMPO_MIDIFILE or (FLUID_PLAYER_TEMPO_INTERNAL) to express that xx_set_tempo() is a function that set the player to follow internal tempo change coming from the MIDI file .

The other enum value ( FLUID_PLAYER_TEMPO_BPM ans FLUID_PLAYER_TEMPO_MIDI) set the player to follow external tempo value provided by xx_set_tempo(). This name should be renamed FLUID_PLAYER_TEMPO_EXTERNAL_BPM and 
FLUID_PLAYER_TEMPO_EXTERNAL_MIDIFILE.

**Comment by jjceresa** [comment]
>Calling get_tempo() with FLUID_PLAYER_TEMPO_BPM or with FLUID_PLAYER_TEMPO_MIDI will give you the value configured by the user with set_tempo() if any, else the default value used by the player (120 bpm). This is consistent with set_tempo(), but not very useful IMO.
But I think that it might be enough to have a simpler API that would just return the current tempo used by the player in the desired format.

To keep consistency with 'xx_set_tempo(),  `xx_get_tempo()` should be able to return external tempo value (FLUID_PLAYER_TEMPO_EXTERNAL_BPM, FLUID_PLAYER_TEMPO_EXTERNAL_MIDIFILE) set by `xx_set_tempo`.
Also `xx_get_tempo()` should be able to return if the player is in internal mode or external mode.

Actually  `xx_get_tempo()` does not return  the current tempo used by the player. This could be added using the enum value
FLUID_PLAYER_TEMPO_CURRENT_BPM and FLUID_PLAYER_TEMPO_CURRENT_MIDIFILE. This will return the tempo actually used by the player in bpm or us/quarter note .

So for xx_get_tempo, I propose the following enum type value:
- FLUID_PLAYER_TEMPO_EXTERNAL_BPM, FLUID_PLAYER_TEMPO_EXTERNAL_MIDIFILE, return external tempo in bpm or us/quarter note.
- FLUID_PLAYER_TEMPO_BPM, FLUID_PLAYER_TEMPO_MIDIFILE, return current tempo used by the player in bpm or us/quarter note.

Opinions ?

**Comment by jjceresa** [comment]
>Could you precise that if no tempo is given in the midi file, a default value of 120 bpm will be used?

Yes, when no tempo change is given in a midi file, a default value of 120 bpm should be used. This is a MIDI file specification.
It will be useful to add this mention in the API comment.

***
>60000000 bpm, really?

Yes, internally valid tempo value should be in the range [1..600000000] regardless if the value is in bpm or in us/quarter note.

**Comment by jjceresa** [comment]
>But I think that it might be enough to have a simpler API that would just return the current tempo used by the player in the desired format.

This is done in recent commit 0836a52 . One just need to call fluid_player_get_tempo() with FLUID_PLAYER_GET_TEMPO_USED_BPM or FLUID_PLAYER_GET_TEMPO_USED_MIDI.

**Comment by derselbst** [comment]
First of all, I'm glad to see that everything is working as expected. Thanks for testing @glm35!

Regarding the `get_tempo()` function: I have a similar feeling as @glm35: It is too overengineered. A client app won't need this. The client app knows whether it has set an absolute tempo or a relative one. Splitting up the enum into a `fluid_player_get_tempo_type` and `fluid_player_set_tempo_type` doesn't really make this simpler. Thinking about it I would rather say to drop `fluid_player_get_tempo()` entirely, due to a missing use case at the present. Instead, the existing functions `fluid_player_get_bpm()` and `fluid_player_get_midi_tempo()` should be continued to use. Those functions obtain the tempo the player is using right now. This might be

* the tempo given by the MIDI file multiplied by the relative factor, or
*  the absolute tempo dictated by the client app.

**Comment by jjceresa** [comment]
>Thinking about it I would rather say to drop fluid_player_get_tempo() entirely, due to a missing use case at the present. Instead, the existing functions fluid_player_get_bpm() and fluid_player_get_midi_tempo() should be continued to use. Those functions obtain the tempo the player is using right now.

Ok.

**Comment by jjceresa** [comment]
Now, the `tempo multiplier` factor set by `fluid_player_set_tempo(..,FLUID_PLAYER_TEMPO_RELATIVE, multiplier)` is only used when the player follow the internal tempo changes. So it seems that we should do the following changes:
1)pass this multiplier value in the `tempo` parameter when calling `fluid_player_set_tempo(..,FLUID_PLAYER_TEMPO_INTERNAL, multiplier)`
2)Get rid of `FLUID_PLAYER_TEMPO_RELATIVE` enum value.

These simplifications allows the caller to set the player in `internal tempo` mode and supplying the `multiplier` factor using only one API function call.
Opinions ?

**Comment by derselbst** [comment]
> Now, the tempo multiplier factor set by fluid_player_set_tempo(..,FLUID_PLAYER_TEMPO_RELATIVE, multiplier) is only used when the player follow the internal tempo changes. So it seems that we should do the following changes:

So, rather than ignoring the `tempo` parameter with `FLUID_PLAYER_TEMPO_INTERNAL`, we use it to pass the `multiplier` directly. Yes, that sounds good to me.

**Review by derselbst — APPROVED** [review]
Looks good now. I'll wait for the CI and then squash it. Thanks!

**Pull request merged**